pkg/pulumiyaml/run.go

```diff
@@ -781,62 +781,69 @@ func (r *Runner) Run(e Evaluator) syntax.Diagnostics {
-		}
-
-		// We have both a default value and a explicit type. Make sure they
-		// agree.
-		if ctypes.IsValidType(expectedType) && t != expectedType {
-			return e.errorf(intm.Key,
-				"type mismatch: default value of type %s but type %s was specified",
-				expectedType, t)
+	var isSecretInConfig bool
+	var defaultValue interface{}
+	var k string
+	var intmKey ast.Expr
+
+	switch intm := intm.(type) {
+	case configNodeYaml:
+		k, intmKey = intm.Key.Value, intm.Key
+		c := intm.Value
+		// If we implement global type checking, the type of configuration variables
+		// can be inferred and this requirement relaxed.
+		if c.Type == nil && c.Default == nil {
+			return e.errorf(intm.Key, "unable to infer type: either 'default' or 'type' is required")
+		}
+		if c.Default != nil {
+			d, ok := e.evaluateExpr(c.Default)
+			if !ok {
+				return nil, false
+			}
+			var err error
+			expectedType, err = ctypes.TypeValue(d)
+			if err != nil {
+				return e.error(c.Default, err.Error())
+			}
+			defaultValue = d
 		}
+		if c.Type != nil {
+			t, ok := ctypes.Parse(c.Type.Value)
+			if !ok {
+				return e.errorf(c.Type,
+					"unexpected configuration type '%s': valid types are %s",
+					c.Type.Value, ctypes.ConfigTypes)
+			}
 
-		expectedType = t
+			// We have both a default value and a explicit type. Make sure they
+			// agree.
+			if ctypes.IsValidType(expectedType) && t != expectedType {
+				return e.errorf(intm.Key,
+					"type mismatch: default value of type %s but type %s was specified",
+					expectedType, t)
+			}
 
-	}
+			expectedType = t
 
-	// A value is considered secret if either it is either marked as secret in
-	// the config section or the configuration section.
-	//
-	// We only want to execute a TrySecret* if the value is secret in the config
-	// section. It the value is specified as secret only in the configuration
-	// section, we call Try* normally, and later wrap the value with
-	// `pulumi.ToSecret`.
-	isSecretInConfig := e.pulumiCtx.IsConfigSecret(e.pulumiCtx.Project() + ":" + k)
+		}
+		// A value is considered secret if either it is either marked as secret in
+		// the config section or the configuration section.
+		//
+		// We only want to execute a TrySecret* if the value is secret in the config
+		// section. It the value is specified as secret only in the configuration
+		// section, we call Try* normally, and later wrap the value with
+		// `pulumi.ToSecret`.
+		isSecretInConfig = e.pulumiCtx.IsConfigSecret(e.pulumiCtx.Project() + ":" + k)
 
-	if isSecretInConfig && c.Secret != nil && !c.Secret.Value {
-		return e.error(c.Secret,
-			"Cannot mark a configuration value as not secret"+
-				" if the associated config value is secret")
+		if isSecretInConfig && c.Secret != nil && !c.Secret.Value {
+			return e.error(c.Secret,
+				"Cannot mark a configuration value as not secret"+
+					" if the associated config value is secret")
+		}
+	case configNodeEnv:
+		k = intm.Key
+		expectedType = intm.Type
```

Comment:
If we never expect to get other node types, we should make sure they don't slip in. If we do, a comment explaining that would be awesome.
```suggestion
		expectedType = intm.Type
	default:
		return e.errorf(nil, "internal error: unexpected config type %T", intm)
```